### PR TITLE
WIP Display popover intelligently

### PR DESCRIPTION
**Pull request by AnshulMalik**
Associated Issue: #2565 

### Summary of Changes

* Added popover direction

### Screenshots
as

### CODE REVIEW

File: src/components/shared/Popover.js [hunk 1]
```diff
@@ -50,12 +58,24 @@ class Popover extends Component {
     const left = targetLeft + targetWidth / 4;
     const top = targetTop - height;
 
-    return { left, top };
+    return { left, top, dir: "up" };
```
Comment (jasonLaster) [inline src/components/shared/Popover.js]: Nice.

We might also want to do left/right here in a bit

File: src/components/shared/Popover.js [hunk 2]
```diff
@@ -15,27 +15,35 @@ class Popover extends Component {
...
     } = this.props.target.getBoundingClientRect();
 
+    let dir = "down";
+
     // width division corresponds to calc in Popover.css
     let left = targetLeft + targetWidth / 2 - width / 5;
     let top = targetBottom;
-    return { left, top };
+    if (top + height > window.innerHeight) {
+      console.log("Going out of bounds");
```
Comment (jasonLaster) [inline src/components/shared/Popover.js]: lets remove this...

Comment (AnshulMalik) [inline src/components/shared/Popover.js]: Oops, I forgot.

File: src/components/shared/Popover.js [hunk 3]
```diff
@@ -50,12 +53,24 @@ class Popover extends Component {
...
-    return { left, top };
+    return { left, top, dir: "up" };
   }
 
   renderPopover() {
     const { children, onMouseLeave } = this.props;
-    const { top, left } = this.state;
+    const { top, left, dir } = this.state;
+
+    if (dir === "up") {
```
Comment (jasonLaster) [inline src/components/shared/Popover.js]: lets see if we can consolidate these two renders into one...

File: src/components/shared/Popover.js [hunk 4]
```diff
@@ -50,12 +53,24 @@ class Popover extends Component {
...
+      return dom.div(
+        {
+          className: "popover popover-up",
```
Comment (jasonLaster) [inline src/components/shared/Popover.js]: we can use the classnames helper to do `classnames("popover", { up: dir =="up" })`

File: src/components/shared/Popover.js [hunk 5]
```diff
@@ -50,12 +53,24 @@ class Popover extends Component {
...
+          onMouseLeave,
+          style: { top, left }
+        },
+        children,
```
Comment (jasonLaster) [inline src/components/shared/Popover.js]: the ordering here is the hard one, but we can likely do something clever :)

File: src/components/shared/Popover.js (hunk 4 above)
Comment (AnshulMalik) [inline src/components/shared/Popover.js]: Didn't know about that. Yeah, we can use this.

**Comment by jasonLaster** [comment]
Hey @AnshulMalik I made some additional follow up tweaks

1. generalized `getChildren` so popovers and tooltips had gaps
2. refactored the CSS so there were popovers and tooltips
3. removed the tooltip caret as the tooltips are much more lightweight interactions 
4. moved the preview function css to its own stylesheet

Let me know what you think and we can land this soon.


Here's a GIF of how it looks now:



I think it's looking pretty good! The only follow up item I see is that we will want to also adjust the horizontal (left/right ) orientation as well. But, lets do this in a different a new issue!

**Comment by AnshulMalik** [comment]
Looks good to me.

**Pull request merged**